src/hotspot/share/runtime/jniHandles.cpp

```diff
@@ -199,13 +199,9 @@ jobjectRefType JNIHandles::handle_type(JavaThread* thread, jobject handle) {
     default:
       ShouldNotReachHere();
     }
-  } else {
+  } else if (is_local_handle(thread, handle) || is_frame_handle(thread, handle)) {
```

Comment:
Should we still add `ShouldNotReachHere()` at global `else` branch? This would make the if-else chain exhaustive with the early warning if some handle type is not used. The prior code did this already.

Reply:
Not sure what you're saying here. As far as I understand the intend of this code is to check whether the handle is of a certain type, and if it's not recognized, return `JNIInvalidRefType`. So, I'm not sure there should be any `ShouldNotReachHere()` in this code.

We can add an `else` branch that returns `JNIInvalidRefType` though.

Reply:
But that just re-introduces the bug? I don't see how we can have both `ShouldNotReachHere()` in the 'else branch' and return `JNIInvalidRefType` for unrecognized handles at the same time.

Not that this function is used to check potentially stale/invalid/corrupt JNI handles, so the input handle can be complete garbage.

Reply:
Ah, d'uh! Not enough coffee today, sorry. Yes, you are right.